### PR TITLE
SPOI-8304: Cassandra sample application

### PR DESCRIPTION
**Pull request by DT-Priyanka**
For details please refer README.md

### CODE REVIEW

File: tutorials/cassandra/README.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+######2. CassandraPojoInputOperator
+
+####Application Flow
+1. TuplesDataGenerator Operator generates test tuples in the form of pojo object (TestUsers)
+2. CassandraDataPopulator Operator writes the test tuples to the cassandra database.
+3. CassandraDataReader Operator reads test tuples from cassandra database. The reader operator gets activated after CassandraDataPopulator is done writing all data to database. CassandraDataPopulator sends a trigger to CassandraDataReader to start read operation.
+4. DataValidator operator receives tuples from CassandraDataReader, it also receives data checksum of input data and does the data validation.
+
+####Configuration
+The config file is bundled with package and is available in src/sit/conf/ folder. Update following properties:
```
Comment (bhupeshchawda) [inline tutorials/cassandra/README.md]: typo: sit => site?

File: tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraApplication.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+    PojoGenerator tuplesGenerator = new PojoGenerator();
+    tuplesGenerator.setTupleNum(tuplesCount);
+
+    CassandraTransactionalStore transactionalStore = new CassandraTransactionalStore();
+
+    CassandraDataPopulator cassandraOutput = new CassandraDataPopulator();
+    cassandraOutput.setStore(transactionalStore);
+    cassandraOutput.setFieldInfos(fieldInfos);
+    cassandraOutput.setNumRowsToWrite(tuplesCount);
```
Comment (bhupeshchawda) [inline tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraApplication.java]: Is this needed even in the original operator?

File: tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraApplication.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+    CassandraInputOperator cassandraInput = new CassandraInputOperator();
+    cassandraInput.setStore(transactionalStore);
+    cassandraInput.setFieldInfos(fieldInfos);
+    cassandraInput.setTuplesCount(tuplesCount);
```
Comment (bhupeshchawda) [inline tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraApplication.java]: Why is this needed?

File: tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraDataPopulator.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+    @Override
+    public void process(TestUsers tuple)
+    {
+      if (writeMoreRecords) {
+        CassandraDataPopulator.super.input.process(tuple);
+        hashCode = hashCode ^ tuple.hashCode();
+        processedTuples++;
+
+        if (processedTuples >= numRowsToWrite) {
```
Comment (bhupeshchawda) [inline tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraDataPopulator.java]: This can be just == ?

File: tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraDataPopulator.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  private int numRowsToWrite;
+  private int processedTuples;
+  private int hashCode;
+  private boolean writeMoreRecords = true;
+
+  @OutputPortFieldAnnotation(optional = true)
+  public transient DefaultOutputPort<ScanInformation> validatationInfo = new DefaultOutputPort<>();
+
+  @InputPortFieldAnnotation(optional = true)
+  public final transient DefaultInputPort<TestUsers> tuplesInput = new DefaultInputPort<TestUsers>()
```
Comment (bhupeshchawda) [inline tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraDataPopulator.java]: I assume this is needed because the process() method of the input port could not be overridden. Can you create a JIRA to abstract out the process method into processTuple() so that this can be called by the extending classes and re-declaration of ports is not needed.

File: tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraDataPopulator.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+  };
+
+  @Override
+  public void setup(OperatorContext context)
+  {
+    Cluster cluster = Cluster.builder().addContactPoint(store.getNode()).build();
+    Session session = cluster.connect(store.getKeyspace());
+    String createMetaTable = "CREATE TABLE IF NOT EXISTS " + store.getKeyspace() + "."
```
Comment (bhupeshchawda) [inline tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraDataPopulator.java]: Is this needed to be done for every operator which uses CassandraOutputOperator? Can this not be done in the original operator?

Comment (bhupeshchawda) [inline tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraDataPopulator.java]: Everything before super.setup()

File: tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraDataPopulator.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        + " (id uuid PRIMARY KEY,fname text,lname text,city text);";
+    session.execute(createTable);
+    session.close();
+    cluster.close();
+
+    super.setup(context);
+  }
+
+  @Override
+  protected Statement setStatementParameters(PreparedStatement updateCommand, Object tuple) throws DriverException
```
Comment (bhupeshchawda) [inline tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraDataPopulator.java]: Is this not done in original operator?

File: tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraInputOperator.java [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+ */
+public class CassandraInputOperator extends CassandraPOJOInputOperator implements IdleTimeHandler
+{
+  private static Logger LOG = LoggerFactory.getLogger(CassandraInputOperator.class);
+  private int tuplesCount;
+  private static int processedTuples;
+  private ScanInformation scanInfo = null;
+  private long scanStartTime;
+  @InputPortFieldAnnotation(optional = true)
+  public transient DefaultInputPort<ScanInformation> scanTrigger = new DefaultInputPort<ScanInformation>()
```
Comment (bhupeshchawda) [inline tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraInputOperator.java]: Why does this need an input port? Can the scan information not be sent to the Validator directly?

File: tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraInputOperator.java [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+  @Override
+  public void emitTuples()
+  {
+    super.emitTuples();
+    try {
+      if (processedTuples >= tuplesCount) {
+        long scanEndTime = System.currentTimeMillis();
+        LOG.info("Time taken to scan " + tuplesCount + " records: " + (scanEndTime - scanStartTime) + " ms");
+        Thread.sleep(1000); //wait till validator gets all tuples
+        validatationInfo.emit(scanInfo);
```
Comment (bhupeshchawda) [inline tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraInputOperator.java]: Again, this can be done at the validator?

File: tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraInputOperator.java [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+    }
+  }
+
+  @Override
+  public void setup(OperatorContext context)
+  {
+    Cluster cluster = Cluster.builder().addContactPoint(store.getNode()).build();
+    Session session = cluster.connect(store.getKeyspace());
+
+    String createMetaTable = "CREATE TABLE IF NOT EXISTS " + store.getKeyspace() + "."
```
Comment (bhupeshchawda) [inline tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraInputOperator.java]: Not done in original operator?

**Comment by bhupeshchawda** [comment]
I had one pass through it. Here is my main concern:
We need to showcase the simplicity in configuring and using the Cassandra Input and Output operators. In this case these are extended from the original operators and some processing is done inside it which makes it difficult for the user to understand the exact configuration needed for the operators. Perhaps the operators shouldn't even be extended, but used directly from malhar. 

I understand that this is a good validation app. But can we use the core operators directly from the library without any modifications? This will enable the user to actually understand how to use these operators in her apps.

File: tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraDataPopulator.java (hunk 6 above)
Comment (DT-Priyanka) [inline tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraDataPopulator.java]: good point, let me check, we don't create meta table in code for other databases as well. Need to give it a thought.

File: tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraDataPopulator.java (hunk 7 above)
Comment (DT-Priyanka) [inline tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraDataPopulator.java]: Need to change the code, after recent changes in Malhar.

File: tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraInputOperator.java (hunk 8 above)
Comment (DT-Priyanka) [inline tutorials/cassandra/src/main/java/com/datatorrent/cassandra/CassandraInputOperator.java]: I wanted this to act as trigger. Cassandra db guaranties eventual consistency, so need the reader operator to wait till writer has done it's job. Let me check if I can make producer send a single to input operator directly.

**Comment by DT-Priyanka** [comment]
I will try to restructure it, to avoid customization. If we don't want customization at all I will have to create two different apps, one for reading and other for writing. The app written for reading, will depend on user to setup database and add data there.
When I wrote this my aim was someone should be able to run the application with minimal setup/configuration time. Also added some extra processing to do validation to showcase the integrity and fault tolerance features.

File: tutorials/README.md [hunk 11]
```diff
@@ -12,3 +12,4 @@
 | operatorTutorial  | Simple example of a word-count operator and a unit test for it |
 | partition         | Shows use of custom partitioning with a stream codec.
 | topnwords         | The Top-N words example. |
+| cassandra         | Shows how to use CassandraPOJOOutputOperator and CassandraPOJOInputOperator in a test application |
```
Comment (amberarrow) [inline tutorials/README.md]: alphabetic order

File: tutorials/cassandra/src/test/java/com/datatorrent/cassandra/CassandraApplicationTest.java [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+  {
+    try {
+      LocalMode lma = LocalMode.newInstance();
+      Configuration conf = new Configuration(false);
+      conf.addResource(this.getClass().getResourceAsStream("/properties-CassandraOutputTestApp.xml"));
+      conf.set("dt.operator.*.prop.store.node", "localhost");
+      lma.prepareDAG(new CassandraApplication(), conf);
+      LocalMode.Controller lc = lma.getController();
+      lc.run(10000); // runs for 10 seconds and quits
+    } catch (ConstraintViolationException e) {
```
Comment (amberarrow) [inline tutorials/cassandra/src/test/java/com/datatorrent/cassandra/CassandraApplicationTest.java]: Please use a test that checks the results as is done in many of the other examples (also noted in the beginners guide).

**Comment by amberarrow** [comment]
Agree with @bhupeshchawda that using the operator directly from the library serves users better. Two different apps is also a good idea since there are likely to be people who are interested in reading but not writing or vice versa. The apps could just read/write from/to HDFS.

For the setup, please take a look at the README.md in jdbcIngest example which provides details instructions as well as scripts for the user on how to setup the DB; something similar could be provided here.

**Comment by DT-Priyanka** [comment]
Will create two different projects, one for cassandra output and other for cassandra input. Closing this pull request and opening new ones.
